model/dao/UserDAO/PostgreSQLUserDAO.php

```diff
@@ -441,64 +451,104 @@ public function update(UserVO $userVO) {
+        try {
+            $statement = $this->pdo->prepare($sql);
+            $statement->bindValue(":login", $userVO->getLogin(), PDO::PARAM_STR);
+            if (!$emptyPassword)
+                $statement->bindValue(":password", $userVO->getPassword(), PDO::PARAM_STR);
+            $statement->execute();
 
-        $affectedRows = pg_affected_rows($res);
+            $userVO->setID($this->pdo->lastInsertId("usr_id_seq"));
 
-        return $affectedRows;
+            $result->setIsSuccessful(true);
+            $result->setMessage('User created successfully.');
+            $result->setResponseCode(201);
+        }
+        catch (PDOException $ex) {
+            $errorMessage = $ex->getMessage();
+            $resultMessage = "Error creating user:\n";
+            if(strpos($errorMessage, "unique_usr_login")) {
+                $resultMessage .= "Login is already used.";
+                $result->setResponseCode(400);
+            }
+            else {
+                $resultMessage .= $errorMessage;
+                $result->setResponseCode(500);
+            }
+            $result->setErrorNumber($ex->getCode());
+            $result->setMessage($resultMessage);
+            $result->setIsSuccessful(false);
+        }
 
+        return $result;
     }
 
     /** User deleter for PostgreSQL.
      *
      * This function deletes the data of a User by its {@link UserVO}.
      *
      * @param UserVO $userVO the {@link UserVO} with the data we want to delete from database.
-     * @return int the number of rows that have been affected (it should be 1).
-     * @throws {@link SQLQueryErrorException}
+     * @return OperationResult the result {@link OperationResult} with information about operation status
      */
     public function delete(UserVO $userVO) {
-        $affectedRows = 0;
+        $result = new OperationResult(false);
 
-        // Check for a user ID.
-        if($userVO->getId() >= 0) {
-            $currUserVO = $this->getById($userVO->getId());
-        }
-
-        // Otherwise delete a user.
-        if(sizeof($currUserVO) > 0) {
-            $sql = "DELETE FROM usr WHERE id=".$userVO->getId();
+        $sql = "DELETE FROM usr WHERE id=:id";
 
-            $res = pg_query($this->connect, $sql);
+        try {
+            $statement = $this->pdo->prepare($sql);
+            $statement->bindValue(":id", $userVO->getId(), PDO::PARAM_INT);
+            $statement->execute();
 
-            if ($res == NULL) throw new SQLQueryErrorException(pg_last_error());
-
-            $affectedRows = pg_affected_rows($res);
+            $result->setIsSuccessful(true);
+            $result->setMessage('User updated successfully.');
+            $result->setResponseCode(201);
+        }
+        catch (PDOException $ex) {
+            $errorMessage = $ex->getMessage();
+            $resultMessage = "Error updating user:\n";
+            if(strpos($errorMessage, "Foreign key violation")) {
+                $result->setResponseCode(403);
```

Comment:
Nit (again, sorry): My understanding of a 403 is that it means a user doesn't have permission to perform the attempted action. A FK violation feels a little more like a generic 500 to me, but I also recognize that some of these response codes don't always fit every situation. 

Same feedback re the message: you might want to specify that that was an error when deleting the user instead of updating.

Reply:
You are right about 403 doesn't really fit, but I don't want to use 500 because it's [defined as a "catch-all" error](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/500) for unexpected situations. It's possible to trigger this foreign key violation from the UI, if an admin attempts to delete a user and that user has already created tasks, so I don't consider it an unexpected situation. And I decided not to use 400 because I don't think it's a client error either...

Maybe [409 Conflict](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409) is a better match?

Reply:
I think 409 makes sense too!